src/hooks/privateChats.ts

```diff
@@ -44,6 +49,60 @@ export const usePrivateChatMessages = () => {
   useConnectPrivateChatMessages();
 
   const privateChatMessages = useSelector(privateChatMessagesSelector);
+  const { user } = useUser();
+  const userId = user?.uid;
+
+  const { chatSettings } = useChatSidebarControls();
+
+  const [play] = useSound(NEW_MESSAGE_SOUND, {
+    // `interrupt` ensures that if the sound starts again before it's
+    // ended, it will truncate it. Otherwise, the sound can overlap.
+    interrupt: true,
+  });
+
+  const [privateChatMessagesState, setPrivateChatMessagesState] = useState(
+    privateChatMessages
+  );
+
+  const chatWithUser =
+    chatSettings.openedChatType === ChatTypes.PRIVATE_CHAT
+      ? chatSettings.recipientId
+      : undefined;
+
+  useEffect(() => {
+    const newMessage = privateChatMessages?.filter(
+      ({ id: id1 }) =>
+        !privateChatMessagesState?.some(({ id: id2 }) => id2 === id1)
+    );
+
+    if (privateChatMessagesState === undefined) {
+      setPrivateChatMessagesState(privateChatMessages);
+    }
+
+    if (
+      privateChatMessages !== undefined &&
+      privateChatMessagesState !== undefined &&
+      newMessage !== undefined &&
+      newMessage.length !== 0 &&
+      privateChatMessages.length !== privateChatMessagesState.length &&
+      !newMessage[0].isRead &&
+      newMessage[0].from !== userId
+    ) {
+      setPrivateChatMessagesState(privateChatMessages);
```

Comment:
I can't be certain off the top of my head as I'm having trouble following the logic here, but this seems way more complex than I would have expected this code needed to be. Is there any way we can simplify it? Or if not, what are the reasons/edge cases that require it to be so verbose?

Reply:
@0xdevalias the logic is first to find which is the new message that the user receives. After that we check if the arrays are empty the first array is the one that comes from the firebase server the other one is the old one that was before the new message. We are also looking for if the new message is also empty and defined. Now we are looking if there is a change between the old array of messages or new array if there is we look if the new message is read if it isn't read we need to see if the message that came from the server is from us or not. In that way we can find out that the user has received message and it isn't from him.
There are 2 more ifs in this one that help us found if the user is in the chat or not, because if he is in the chat there should not be notification playing. So to sum up this if helps us find if there is a new message and if the user is in the chat or not.